agent/src/main/java/com/teamscale/jacoco/agent/options/ClasspathUtils.java

```diff
@@ -14,25 +14,33 @@
 
 	/** Replaces all txt files in the given list with the file names written in the txt file separated by new lines. */
 	public static List<File> resolveClasspathTextFiles(String key, FilePatternResolver filePatternResolver,
-													   List<String> classDirectoriesOrZips) throws AgentOptionParseException {
-		List<File> classDirectoriesOrZipFiles = new ArrayList<>();
-		for (String classDirectoryOrZip : classDirectoriesOrZips) {
-			classDirectoriesOrZipFiles.addAll(filePatternResolver.resolveToMultipleFiles(key, classDirectoryOrZip));
+													   List<String> patterns) throws AgentOptionParseException {
+		List<File> resolvedPaths = new ArrayList<>();
+		for (String pattern : patterns) {
+			resolvedPaths.addAll(filePatternResolver.resolveToMultipleFiles(key, pattern));
 		}
-		Map<Boolean, List<File>> filesByType = classDirectoriesOrZipFiles.stream()
-				.collect(Collectors.partitioningBy(fileName -> fileName.getName().endsWith(".txt")));
-		List<File> classJarOrDirFiles = new ArrayList<>(filesByType.get(false));
+		Map<Boolean, List<File>> filesByType = resolvedPaths.stream()
+				.collect(Collectors.partitioningBy(file -> file.getName().endsWith(".txt")));
+		List<File> classDirOrJarFiles = new ArrayList<>(filesByType.get(false));
 		for (File txtFile : filesByType.get(true)) {
-			try {
-				for (String file : FileSystemUtils.readLinesUTF8(txtFile)) {
-					classJarOrDirFiles.addAll(filePatternResolver.resolveToMultipleFiles(key, file));
-				}
-			} catch (IOException e) {
-				throw new AgentOptionParseException("Failed to resolve class path entries from the text files provided " +
-						"in the `" + key + "` option. Please ensure that any text files provided in this option contain " +
-						"properly formatted classpath pattern and each entry is separated by a newline.", e);
+			classDirOrJarFiles.addAll(resolveClassPathEntries(key, filePatternResolver, txtFile));
+		}
+		return classDirOrJarFiles;
+	}
+
+	private static List<File> resolveClassPathEntries(String key, FilePatternResolver filePatternResolver,
+													  File txtFile) throws AgentOptionParseException {
+		try {
+			List<String> filePaths = FileSystemUtils.readLinesUTF8(txtFile);
+			List<File> resolvedFiles = new ArrayList<>();
+			for (String filePath : filePaths) {
+				resolvedFiles.addAll(filePatternResolver.resolveToMultipleFiles(key, filePath));
 			}
+			return resolvedFiles;
+		} catch (IOException e) {
+			throw new AgentOptionParseException("Failed to resolve class path entries from the text files provided " +
```

Comment:
AFAICS from the code, `IOException` can only be thrown by `FileSystemUtils.readLinesUTF8`, meaning that it's not a problem with the content of the file (e.g. newline separation) but rather with e.g. permissions or other file system errors. Could you please change the exception message to reflect this, e.g. advise the user to check the permissions of the file that caused the problem?

In fact, I thnik it would help the readability of the method if we constrained the try-catch to only the FileSystemUtils call - then it's immediately obvious that that's the case the catch handles :-)

Reply:
👍